tests/test_train_taskcluster.py

```diff
@@ -76,8 +80,262 @@
     ),
 )
 def test_all_args_forwarded(args):
-    with mock.patch("train_taskcluster.subprocess") as mocked_subprocess:
-        train_taskcluster.main(args)
-        assert mocked_subprocess.run.call_args_list == [
-            mock.call([TRAIN_TASKCLUSTER_SH] + args, check=True),
-        ]
+    with mock.patch.multiple(
+        train_taskcluster, subprocess=mock.DEFAULT, requests=mock.DEFAULT
+    ) as tt_mock:
+        with mock.patch.dict(os.environ) as mocked_env:
+            mocked_env["TASK_ID"] = "abcdef"
+            mocked_env["RUN_ID"] = "0"
+            mocked_env["TASKCLUSTER_ROOT_URL"] = "https://some.cluster"
+            train_taskcluster.main(args)
+            assert tt_mock["subprocess"].run.call_args_list == [
+                mock.call([TRAIN_TASKCLUSTER_SH] + args, check=True),
+            ]
+
+
+FULL_ARTIFACTS = [
+    {
+        "storageType": "s3",
+        "name": f"public/build/{artifact}",
+        "expires": "2035-04-23T16:22:13.477Z",
+        "contentType": "application/x-yaml",
+    }
+    # Note: we are depending on something under to test for construction of parts of
+    # these tests. However, the value of duplicating this list into the tests instead
+    # is debatable, as what we're _really_ testing is the logic around them, not
+    # which ones are or are not required. It seems reasonable to keep a single source
+    # of truth for this.
+    for artifact in train_taskcluster.CONTINUATION_ARTIFACTS
+]
+FULL_ARTIFACTS.append(
+    {
+        "storageType": "s3",
+        "name": "public/logs/live.log",
+        "expires": "2035-04-23T16:22:13.477Z",
+        "contentType": "text/plain; charset=utf-8",
+    },
+)
+PARTIAL_ARTIFACTS = [
+    {
+        "storageType": "s3",
+        "name": "public/build/model.npz",
+        "expires": "2035-04-23T16:22:13.477Z",
+        "contentType": "application/x-yaml",
+    },
+    {
+        "storageType": "s3",
+        "name": "public/logs/live.log",
+        "expires": "2035-04-23T16:22:13.477Z",
+        "contentType": "text/plain; charset=utf-8",
+    },
+]
+
+
+@pytest.mark.parametrize(
+    "current_run_id,resumable_run_id,reasons_created,run_artifacts,orig_pretrained_model_mode,expected_pretrained_model_mode",
```

Comment:
Thought (no action need): This style of parameterization is hard to connect the parameters with their names. I'm not sure if there is a way to label them easier with a TypedDict or something. I do like the fact that they have `id`s on them.

Reply:
I agree. I'm not aware of any way to do this with [pytest params](https://docs.pytest.org/en/latest/example/parametrize.html) :(. It looks like https://github.com/pytest-dev/pytest/issues/9216 talks about this a bit.